x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,367 @@
+                return this;
+            }
+            return textExpansionResultsToQuery(fieldName, vectorDimensionsSupplier.get());
+        }
+
+        if (modelId != null) {
+            CoordinatedInferenceAction.Request inferRequest = CoordinatedInferenceAction.Request.forTextInput(
+                modelId,
+                List.of(modelText),
+                TextExpansionConfigUpdate.EMPTY_UPDATE,
+                false,
+                InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API
+            );
+            inferRequest.setHighPriority(true);
+            inferRequest.setPrefixType(TrainedModelPrefixStrings.PrefixType.SEARCH);
+
+            SetOnce<TextExpansionResults> textExpansionResultsSupplier = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                (client, listener) -> executeAsyncWithOrigin(
+                    client,
+                    ML_ORIGIN,
+                    CoordinatedInferenceAction.INSTANCE,
+                    inferRequest,
+                    ActionListener.wrap(inferenceResponse -> {
+
+                        if (inferenceResponse.getInferenceResults().isEmpty()) {
+                            listener.onFailure(new IllegalStateException("inference response contain no results"));
+                            return;
+                        }
+
+                        if (inferenceResponse.getInferenceResults().get(0) instanceof TextExpansionResults textExpansionResults) {
+                            textExpansionResultsSupplier.set(textExpansionResults);
+                            listener.onResponse(null);
+                        } else if (inferenceResponse.getInferenceResults().get(0) instanceof WarningInferenceResults warning) {
+                            listener.onFailure(new IllegalStateException(warning.getWarning()));
+                        } else {
+                            listener.onFailure(
+                                new IllegalStateException(
+                                    "expected a result of type ["
+                                        + TextExpansionResults.NAME
+                                        + "] received ["
+                                        + inferenceResponse.getInferenceResults().get(0).getWriteableName()
+                                        + "]. Is ["
+                                        + modelId
+                                        + "] a compatible model?"
+                                )
+                            );
+                        }
+                    }, listener::onFailure)
+                )
+            );
+
+            return new SparseVectorQueryBuilder(this, textExpansionResultsSupplier);
+
+        } else {
+            return vectorDimensionsToQuery(fieldName, vectorDimensions);
+        }
+
+    }
+
+    private QueryBuilder vectorDimensionsToQuery(String fieldName, List<VectorDimension> vectorDimensions) {
```

Comment:
Since this is a new query, this should be able to always use the vector dimensions query builder (previously weighted tokens query builder). However, I'm getting CCS validation errors locally when trying to do this. There might be something I'm missing here, but it would make it much cleaner if I could do away with this logic.

Reply:
I don't think this rewrite should occur at all.

`SparseVectorQueryBuilder` accepts a `query_vector` or can have the `query_vector` created automatically given a `model_id` and `text`.

There is no need to rewrite this as it has all the information it needs even if the pruning config is null.